app/node/distort/cornerpin/cornerpindistortnode.cpp

```diff
@@ -0,0 +1,237 @@
+
+                                  top_left.x(), top_left.y(), 0.0f,
+                                  bottom_left.x(),  bottom_left.y(), 0.0f,
+                                  bottom_right.x(),  bottom_right.y(), 0.0f};
+  job.SetVertexCoordinates(adjusted_vertices);
+
+  // If no texture do nothing
+  if (!job.GetValue(kTextureInput).data().isNull()) {
+    // In the special case that all sliders are in their default position just
+    // push the texture.
+    if (!(job.GetValue(kTopLeftInput).data().value<QVector2D>().isNull()
+        && job.GetValue(kTopRightInput).data().value<QVector2D>().isNull() &&
+        job.GetValue(kBottomRightInput).data().value<QVector2D>().isNull() &&
+        job.GetValue(kBottomLeftInput).data().value<QVector2D>().isNull())) {
+      table->Push(NodeValue::kShaderJob, QVariant::fromValue(job), this);
+    } else {
+      table->Push(NodeValue::kTexture, job.GetValue(kTextureInput).data(), this);
+    }
+  }
+}
+
+ShaderCode CornerPinDistortNode::GetShaderCode(const QString &shader_id) const
+{
+  Q_UNUSED(shader_id)
+  return ShaderCode(FileFunctions::ReadFileAsString(QStringLiteral(":/shaders/cornerpin.frag")),
+                    FileFunctions::ReadFileAsString(QStringLiteral(":/shaders/cornerpin.vert")));
+}
+
+QPointF CornerPinDistortNode::ValueToPixel(int value, const NodeValueRow& row, const QVector2D &resolution) const
+{
+  Q_ASSERT(value >= 0 && value <= 3);
+  switch (value) {
+    case 0: // Top left
+      return QPointF(row[kTopLeftInput].data().value<QVector2D>().x(),
+                     row[kTopLeftInput].data().value<QVector2D>().y());
+      break;
+    case 1: // Top right
+      return QPointF(resolution.x() + row[kTopRightInput].data().value<QVector2D>().x(),
+                     row[kTopRightInput].data().value<QVector2D>().y());
+      break;
+    case 2: // Bottom right
+      return QPointF(resolution.x() + row[kBottomRightInput].data().value<QVector2D>().x(),
+                     resolution.y() + row[kBottomRightInput].data().value<QVector2D>().y());
+      break;
+    case 3: //Bottom left
+      return QPointF(row[kBottomLeftInput].data().value<QVector2D>().x(),
+                     row[kBottomLeftInput].data().value<QVector2D>().y() + resolution.y());
+      break;
+    default: // We should never get here
+      return QPointF();
+  }
+}
+
+void CornerPinDistortNode::DrawGizmos(const NodeValueRow &row, const NodeGlobals &globals, QPainter *p)
+{
+  const QVector2D &resolution = globals.resolution();
+
+  const double handle_radius = GetGizmoHandleRadius(p->transform());
+
+  p->setPen(QPen(Qt::white, 0));
+
+  QPointF top_left = ValueToPixel(0, row, resolution);
+  QPointF top_right = ValueToPixel(1, row, resolution);
+  QPointF bottom_right = ValueToPixel(2, row, resolution);
+  QPointF bottom_left = ValueToPixel(3, row, resolution);
+
+  // Add the correct offset to each slider
+  SetInputProperty(kTopLeftInput, QStringLiteral("offset"), QVector2D(0.0, 0.0));
```

Comment:
It might be good if `ValueToPixel` could be used to calculate the offsets too. It would mean ValueToPixel would have to take a QVector2D instead of a NodeValueRow, but you could have an additional override that pulls the QVector2D from the row so both usages are valid. e.g:

```
QPointF CornerPinDistortNode::ValueToPixel(int value, const NodeValueRow& row, const QVector2D &resolution) const
{
  QString column;
  switch (value) {
  case 0:
    column = kTopLeftInput;
    break;
  etc. etc.
  }

  return ValueToPixel(value, row[column].data().value<QVector2D>(), resolution);
}
```

Reply:
Sorry I'm not quite sure what you mean here.